src/protojure/pedestal/core.clj

```diff
@@ -241,7 +244,12 @@
 (defn unsubscribe-close [connections ^HttpServerExchange exchange]
   (let [conn (.getConnection exchange)
         k (hash conn)]
-    (handle-disconnect connections k)))
+    (swap! connections
+           (fn [x]
+             (update x k (fn [y]
+                           (let [channels (get y exchange)]
```

Comment:
The one thing that strikes me as a little odd here is the use of the naked exchange object vs something like creating a k2 (hash exchange) to use -- was there any reason to explicitly has the connection, but not the exchange prior to use as a key?

Reply:
Im not sure that it matters, but I agree I should be consistent.  I added hashes to the connection tracking because I misdiagnosed a problem as not being able to find the connection in the map when it was in fact something else.  Ill clean this up so its the same.

Reply:
I'm not sure what behavior get has when an object is passed as a key -- it could be calling .hashcode or it could be using the memory location reference, or something else -- given some of the performance scenarios you are seeing over the life time of a connection, it may make sense to be defensive in avoiding collisions and explicitly hashing as you did -- but I'm also not sure that this is a real issue (I tried cursorily digging into the backing impl of get, but the source required a second hop deeper into the reader internals than I cared for.

Reply:
I was actually worried about the opposite: collisions in hash-code, vs the specificity of a pointer key.  As far as I can tell, this is legitimate, and here is a moderate proof.

```
(new java.lang.Object)
=> #object[java.lang.Object 0xf191678 "java.lang.Object@f191678"]
(def x (new java.lang.Object))
=> #'user/x
(def y (new java.lang.Object))
=> #'user/y
(get {x "foo" y "bar"} x)
=> "foo"
(get {x "foo" y "bar"} y)
=> "bar"
```

Reply:
Yeah, this is what I was pointing out --

```
user=> (def x (new java.lang.Object))
#'user/x
user=> (def y (new java.lang.Object))
#'user/y
user=> (def m {x "foo" y "bar"})
#'user/m
 user=> (keys m)
(#object[java.lang.Object 0x7ab8c920 "java.lang.Object@7ab8c920"] #object[java.lang.Object 0x6842ca0e "java.lang.Object@6842ca0e"])
```

The jvm has two basic ways of checking equality -- normally for objects its .equals, which will delegate to comparing:
```
user=> (.hashCode x)
2058930464
user=> (.hashCode y)
1749207566
user=> (= (.hashCode x) (.hashCode y))
false
```

This can be tricky because of:
>  It is not required that if two objects are unequal according to the equals(java.lang.Object) method, then calling the hashCode method on each of the two objects must produce distinct integer results. However, the programmer should be aware that producing distinct integer results for unequal objects may improve the performance of hash tables.

and

> (This is typically implemented by converting the internal address of the object into an integer, but this implementation technique is not required by the JavaTM programming language.)

https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode()

Particularly since things on the Exchange object may change, which under some implementations, means the hashcode can change.

You're _probably_ ok with this implementation 80% of the time, but it would be good to know 1) what does get call on an object to use it as a key, 2) if it uses .hashCode, how does the Exchange object implement it, 3) if the exchange object uses memory location, are there scenarios where an Exchange object may be remapped to a new memory address while it exists in the connections atom -- because this may be an issue.

But point taken that knowing what to use from the Exchange object for uniqueness is the tricky part here.